File: go/e2e/blob_local_test.go
//go:build e2e

package e2e

import (
	"bytes"
	"context"
	"strings"
	"testing"
	"time"

	goystore "github.com/goy-co/goy-store/go"
)

func TestLocalBlobStore_Operations(t *testing.T) {
	store := createTestStore(t)
	ctx := context.Background()

	key := "configs/node-1/app.json"
	payload := bytes.Repeat([]byte{0xAB}, 64*1024) // 64KB
	meta := &goystore.Metadata{
		ContentType: "application/json",
		Custom: map[string]string{
			"author": "goy-team",
		},
	}

	// 1. Put
	if err := store.Blob().Put(ctx, key, payload, meta); err != nil {
		t.Fatalf("Put failed: %v", err)
	}

	// 2. Get
	fetched, fetchedMeta, exists, err := store.Blob().Get(ctx, key)
	if err != nil || !exists {
		t.Fatalf("Get failed: exists=%v, err=%v", exists, err)
	}
	if !bytes.Equal(fetched, payload) {
		t.Fatalf("fetched data does not match payload")
	}
	if fetchedMeta.ContentType != "application/json" || fetchedMeta.Custom["author"] != "goy-team" {
		t.Fatalf("metadata mismatch: %+v", fetchedMeta)
	}

	// 3. List
	prefix := "configs"
	list, err := store.Blob().List(ctx, &prefix)
	if err != nil || len(list) != 1 || list[0] != key {
		t.Fatalf("List failed: list=%+v, err=%v", list, err)
	}

	// 4. PresignURL
	url, err := store.Blob().PresignURL(ctx, key, time.Minute)
	if err != nil || !strings.HasPrefix(url, "file://") {
		t.Fatalf("PresignURL failed: url=%s, err=%v", url, err)
	}

	// 5. Delete
	if err := store.Blob().Delete(ctx, key); err != nil {
		t.Fatalf("Delete failed: %v", err)
	}

	_, _, exists, _ = store.Blob().Get(ctx, key)
	if exists {
		t.Fatalf("blob should not exist after deletion")
	}
}

File: rust/tests/blob_local_test.rs
use super::common::*;
use goy_store::blob::Metadata;
use std::collections::HashMap;
use std::time::Duration;

#[tokio::test]
async fn test_local_blob_store_operations() {
    let store = create_test_store().await.expect("failed to create store");

    let key = "configs/node-1/app.json";
    let payload = vec![0xAB; 64 * 1024]; // 64KB
    let mut custom = HashMap::new();
    custom.insert("author".to_string(), "goy-team".to_string());

    let meta = Metadata {
        content_type: Some("application/json".to_string()),
        custom,
    };

    // 1. Put
    store
        .blob
        .put(key, &payload, Some(meta.clone()))
        .await
        .unwrap();

    // 2. Get
    let fetched = store.blob.get(key).await.unwrap();
    assert!(fetched.is_some());
    let (data, fetched_meta) = fetched.unwrap();
    assert_eq!(data, payload);
    assert_eq!(
        fetched_meta.content_type,
        Some("application/json".to_string())
    );
    assert_eq!(fetched_meta.custom.get("author").unwrap(), "goy-team");

    // 3. List with prefix
    let list = store.blob.list(Some("configs")).await.unwrap();
    assert_eq!(list, vec!["configs/node-1/app.json"]);

    // 4. Presign URL
    let url = store
        .blob
        .presign_url(key, Duration::from_secs(60))
        .await
        .unwrap();
    assert!(url.starts_with("file://"));

    // 5. Delete
    store.blob.delete(key).await.unwrap();
    let after_del = store.blob.get(key).await.unwrap();
    assert!(after_del.is_none());
}
